src/Symfony/Component/Cache/Adapter/RecordingAdapter.php

```diff
@@ -0,0 +1,202 @@
+ * @author Aaron Scherer <aequasi@gmail.com>
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class RecordingAdapter implements AdapterInterface
+{
+    /**
+     * @var array
+     */
+    private $calls = array();
+
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cachePool;
+
+    /**
+     * @var Stopwatch
+     */
+    private $stopwatch;
+
+    /**
+     * @param CacheItemPoolInterface $cachePool
+     * @param Stopwatch              $stopwatch
+     */
+    public function __construct(CacheItemPoolInterface $cachePool, Stopwatch $stopwatch)
+    {
+        $this->cachePool = $cachePool;
+        $this->stopwatch = $stopwatch;
+    }
+
+    /**
+     * @param string $name
+     * @param array  $arguments
+     *
+     * @return object
+     */
+    private function timeCall($name, array $arguments = array())
+    {
+        $time = 0;
+        $event = $this->stopwatch->start(get_class($this->cachePool), 'cache');
+        $result = call_user_func_array(array($this->cachePool, $name), $arguments);
+        if ($event->isStarted()) {
+            $event->stop();
+            $time = $event->getEndTime() - $event->getStartTime();
+        }
+
+        $object = (object) compact('name', 'arguments', 'start', 'time', 'result');
+
+        return $object;
+    }
+
+    public function getItem($key)
+    {
+        $call = $this->timeCall(__FUNCTION__, array($key));
+        $result = $call->result;
+        $call->isHit = $result->isHit();
+
+        // Display the result in a good way depending on the data type
+        if ($call->isHit) {
+            $call->result = $this->getValueRepresentation($result);
+        } else {
+            $call->result = null;
+        }
+
+        $this->calls[] = $call;
+
+        return $result;
+    }
+
+    public function hasItem($key)
+    {
+        $call = $this->timeCall(__FUNCTION__, array($key));
+        $this->calls[] = $call;
+
+        return $call->result;
+    }
+
+    public function deleteItem($key)
+    {
+        $call = $this->timeCall(__FUNCTION__, array($key));
+        $this->calls[] = $call;
+
+        return $call->result;
+    }
+
+    public function save(CacheItemInterface $item)
+    {
+        $arg = $this->getValueRepresentation($item);
+
+        $call = $this->timeCall(__FUNCTION__, array($item));
+        $call->arguments = array($arg);
+        $this->calls[] = $call;
+
+        return $call->result;
+    }
+
+    public function saveDeferred(CacheItemInterface $item)
+    {
+        $arg = $this->getValueRepresentation($item);
+
+        $call = $this->timeCall(__FUNCTION__, array($item));
+        $call->arguments = array($arg);
+        $this->calls[] = $call;
+
+        return $call->result;
+    }
+
+    public function getItems(array $keys = array())
+    {
+        $call = $this->timeCall(__FUNCTION__, array($keys));
+        $result = $call->result;
+        $hits = 0;
+        $items = array();
+        foreach ($result as $item) {
```

Comment:
In the case of a generator, I see two issues here:
- this is consuming the generator. Since these are not rewindable, `return $result;` returns a non usable generator.
- this is couting hits, but maybe the app didn't consume the generator on its side.

This means I think we should have two code paths here: one for array (doing the same as the current code),
and one for Traversable, which should wrap this in a generator-closure, and do things lazily.